File: weight_methods.py
import math
from abc import ABC, abstractmethod
import numpy as np
from scipy.special import comb


class WeightingTemplate(ABC):
    """
    Blueprint for weighting mechanisms.
    """

    @abstractmethod
    def compute_weight(self, E):
        """
        Compute the marginal ship weight from an integer summary.

        Parameters:
        E: integer

        Returns:
        unnormalized marginal weight
        """
        pass

    @abstractmethod
    def compute_energy(self, intersection):
        """
        Compute the ship's energy given its intersection with the current board.

        Parameters:
        intersection: view of grid squares

        Returns:
        energy
        """
        pass

    @abstractmethod
    def update_energy(self, E, S, D):
        """
        Update the energy given a change at a single square.

        Parameters:
        E: current energy
        S: current square value
        D: change in square value

        return:
        energy
        """
        # Later logic relies on D not being 0. I shouldn't call update if there is no update.
        pass

    @abstractmethod
    def compute_square_energy(self, arr):
        """
        Performs elementwise the energy contribution of a np array.

        Parameters:
        arr: np.array
        """
        pass


class HardLattice(WeightingTemplate):
    """
    Weighting associated with assigning equal probability to all zero
    intersection states. A ship has weight 1 or 0 depending on whether it can be
    placed on the current board.
    """

    def compute_weight(self, E):
        return int(E == 0)

    def compute_energy(self, intersection):
        return sum(intersection)

    def update_energy(self, E, S, D):
        return E + D

    def compute_square_energy(self, arr):
        return np.zeros_like(arr)


class BinaryBoltzmann(WeightingTemplate):
    """
    Weighting associated with max(0, min(1, n-1)) energy per square. The ship
    contributes 1 unit of energy for each square where occupancy was 1 before
    placement of this ship.
    """

    def __init__(self, beta):
        self.beta = beta

    def compute_weight(self, E):
        return math.exp(-self.beta * E)

    def compute_energy(self, intersection):
        return sum(x == 1 for x in intersection)

    def update_energy(self, E, S, D):
        if S == 1:
            return E + 1
        elif S - D == 1:
            return E - 1
        else:
            return E

    def compute_square_energy(self, arr):
        return np.maximum(0, np.minimum(1, arr - 1))


class PairwiseBoltzmann(WeightingTemplate):
    """
    Weighting associated with n choose 2 energy per square with n occupants. The
    ship contributes n units of energy for each square with n occupants.
    """

    def __init__(self, beta):
        self.beta = beta

    def compute_weight(self, E):
        return math.exp(-self.beta * E)

    def compute_energy(self, intersection):
        return sum(intersection)

    def update_energy(self, E, S, D):
        return E + D

    def compute_square_energy(self, arr):
        return comb(arr, 2)


class OverflowBoltzmann(WeightingTemplate):
    """
    Weighting associated with max(0, n-1) energy per square with n occupants. The
    ship contributes 1 units of energy for each square with 1 or more occupants.
    """

    def __init__(self, beta):
        self.beta = beta

    def compute_weight(self, E):
        return math.exp(-self.beta * E)

    def compute_energy(self, intersection):
        return sum([x > 0 for x in intersection])

    def update_energy(self, E, S, D):
        if S == 0:
            return E - 1
        elif S == 1 and D == 1:
            return E + 1
        else:
            return E

    def compute_square_energy(self, arr):
        return np.maximum(0, arr - 1)

File: CVersion/WeightMethods.cpp
#include "WeightMethods.h"
#include <cmath>

using namespace std;

// HardLattice Implementation
float HardLattice::compute_weight(int E) const
{
    if (E == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

int HardLattice::compute_energy(vector<int> &intersection) const
{
    int E = 0;
    for (int i = 0; i < intersection.size(); i++)
    {
        E += intersection[i];
    }
    return E;
}

int HardLattice::compute_square_energy(int s) const
{
    return 0;
}

int HardLattice::update_energy(int E, int S, int D) const
{
    return E + D;
}

int HardLattice::max_energy(int k, int q) const
{
    return 1;
}

// BinaryBoltzmann implementation
BinaryBoltzmann::BinaryBoltzmann(float beta)
{
    this->beta = beta;
}

float BinaryBoltzmann::compute_weight(int E) const
{
    return exp(-beta * E);
}

int BinaryBoltzmann::compute_energy(vector<int> &intersection) const
{
    int E = 0;
    for (int i = 0; i < intersection.size(); i++)
    {
        if (intersection[i] == 1)
        {
            E++;
        }
    }
    return E;
}

int BinaryBoltzmann::compute_square_energy(int s) const
{
    return max(0, min(1, s - 1));
}

int BinaryBoltzmann::update_energy(int E, int S, int D) const
{
    if (S == 1)
    {
        E++;
        return E;
    }
    else if ((S - D) == 1)
    {
        E--;
        return E;
    }
    else
    {
        return E;
    }
}

int BinaryBoltzmann::max_energy(int k, int q) const
{
    return k;
}

// PairwiseBoltzmann implementation
PairwiseBoltzmann::PairwiseBoltzmann(float beta)
{
    this->beta = beta;
}

float PairwiseBoltzmann::compute_weight(int E) const
{
    return exp(-beta * E);
}

int PairwiseBoltzmann::compute_energy(vector<int> &intersection) const
{
    int E = 0;
    for (int i = 0; i < intersection.size(); i++)
    {
        E += intersection[i];
    }
    return E;
}

int PairwiseBoltzmann::update_energy(int E, int S, int D) const
{
    return E + D;
}

int PairwiseBoltzmann::compute_square_energy(int s) const
{
    return s * (s - 1) / 2;
}

int PairwiseBoltzmann::max_energy(int k, int q) const
{
    return k * (q - 1);
}

// OverflowBoltzmann implementation
OverflowBoltzmann::OverflowBoltzmann(float beta)
{
    this->beta = beta;
}

float OverflowBoltzmann::compute_weight(int E) const
{
    return exp(-beta * E);
}

int OverflowBoltzmann::compute_energy(vector<int> &intersections) const
{
    int E = 0;
    for (int i = 0; i < intersections.size(); i++)
    {
        if (intersections[i] > 0)
        {
            E++;
        }
    }
    return E;
}

int OverflowBoltzmann::update_energy(int E, int S, int D) const
{
    if (S == 0)
    {
        return E - 1;
    }
    else if (S == 1 && D == 1)
    {
        return E + 1;
    }
    else
    {
        return E;
    }
}

int OverflowBoltzmann::compute_square_energy(int s) const
{
    return max(0, s - 1);
}

int OverflowBoltzmann::max_energy(int k, int q) const
{
    return k;
}